cmd/databroker_options.go

```diff
@@ -0,0 +1,74 @@
+package cmd
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/spf13/pflag"
+	"github.com/volatiletech/null/v9"
+	"google.golang.org/grpc"
+	"gopkg.in/yaml.v3"
+
+	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/pkg/grpcutil"
+)
+
+type dataBrokerOptions struct {
+	ClusterNodeID   string
+	ClusterNodes    dataBrokerClusterNodes
+	RaftBindAddress string   `validate:"omitempty,hostname_port"`
+	ServiceURLs     []string `validate:"dive,url"`
+}
+
+// dataBrokerClusterNodes is a custom type to support reading cluster nodes from yaml
+type dataBrokerClusterNodes []config.DataBrokerClusterNode
+
+func (n dataBrokerClusterNodes) MarshalText() ([]byte, error) {
+	return yaml.Marshal([]config.DataBrokerClusterNode(n))
+}
+
+func (n *dataBrokerClusterNodes) UnmarshalText(text []byte) error {
+	return yaml.Unmarshal(text, (*[]config.DataBrokerClusterNode)(n))
+}
+
+func (o *dataBrokerOptions) setupFlags(flags *pflag.FlagSet) {
+	flags.StringVar(&o.ClusterNodeID, "databroker-cluster-node-id", "", "the databroker cluster node id")
+	flags.TextVar(&o.ClusterNodes, "databroker-cluster-nodes", dataBrokerClusterNodes(nil), "the databroker cluster nodes")
+	flags.StringVar(&o.RaftBindAddress, "databroker-raft-bind-address", "", "the databroker raft bind address")
+	flags.StringSliceVar(&o.ServiceURLs, "databroker-service-urls", nil, "the databroker service urls, defaults to localhost")
+}
+
+func (o *dataBrokerOptions) apply(dst *config.Config) {
+	if o.ClusterNodeID != "" {
+		dst.Options.DataBroker.ClusterNodeID = null.StringFrom(o.ClusterNodeID)
+	}
+	dst.Options.DataBroker.ClusterNodes = config.DataBrokerClusterNodes(o.ClusterNodes)
+	if o.RaftBindAddress != "" {
+		dst.Options.DataBroker.RaftBindAddress = null.StringFrom(o.RaftBindAddress)
+	}
+	dst.Options.DataBroker.ServiceURLs = o.ServiceURLs
+}
+
+func getDataBrokerConnection(ctx context.Context, cfg *config.Config) (*grpc.ClientConn, error) {
```

Comment:
there's an overlap with https://github.com/pomerium/ingress-controller/blob/9a691e4df7eef2348f4bdc22870140e2ef880aaf/cmd/controller.go#L109

is it possible to unify them into a single function?